### DISCUSSION TITLE
Islands Architecture

### DISCUSSION THREAD

**Discussion by kvnwolf** [remix-run/remix#3861] [Proposals]
## What is the Islands Architecture?

The idea of islands (currently used by some SSR frameworks like [Astro](https://docs.astro.build/en/concepts/islands/) and [Fresh](https://fresh.deno.dev/docs/concepts/islands)). When using the islands architecture/pattern, the server takes care of only rendering and sending plain HTML, while the client takes charge of hydrating the island component.

Take the following image as an example:

![image](https://user-images.githubusercontent.com/3157426/181640739-d726d67b-3545-4b5f-8dd4-9043d0874903.png)

On the above image, the `Sidebar`, `Header`,  and `Image Carousel` components are parts of the application that needs to be interactive, but not immediately, only after the code for them is loaded.

## How can this work with Remix?

Currently, the only way to decide whether to render - or not render at all - the `<script />` tags pointing to the route `chunks`, is to render or not render the `<Scripts />` component exported by `@remix-run/react`.

My idea/proposal (although I haven't dug that much into the Remix compiler) is to support the inclusion of `*.island.tsx?` files. This way, it could try to create a dependency graph, extract chunks of shared code into modules and add the definitions to the `remixManifest`.

Again, I haven't dug that deep into how the compiler, or module resolution works, but if this gets merged into Remix, then it could evolve to eventually deprecate the `<Scripts />` tag, and always sending zero JavaScript to the client, unless an island import is detected.

**Comment by sergiodxa** [comment]
If you move the interactivity of your UI outside the React components used to render React server-side, you could then load your own client-side JS code from the public folder only on the route you need them.

This way you can add interactivity only for the parts you need.

**Reply by kvnwolf** [reply]
Yes, I know that can be a way to do it. However, I might not be able to consume the code of my app, nor use TypeScript. With the proposed architecture, you will be able to select which components you might want to hydrate/add scripts to the client. That way, you could, for example, have a `<Video />` component (like the youtube one, with all the interactivity) only loading the code it needs, and if later there is another component you might want to be interactive, for instance, a `<LiveChat />`, following the YouTube example on live videos, then the compiler could determine which chunks are shared by them, and send them to the browser together with the component code.

**Reply by sergiodxa** [reply]
Well as long as Remix uses React you will have to wait for React Server Components to do that, because otherwise React needs to control everything

**Reply by kvnwolf** [reply]
Yes, I know React needs to control anything, but take for example [Fresh](https://fresh.deno.dev), they can send only the scripts required by the interactive islands to the browser.

I think that maybe the right term could be "selective hydration", where I decide which components will be hydrated on the client, and only render script tags for the code that component depends on.

**Comment by hollandThomas** [comment]
Full disclosure: I haven't found the time to toy around with this yet at all, it's just a tab that's been sitting there for a while. That being said, Podium looks like it might be the thing you're looking for
https://podium-lib.io/

**Comment by dangayle** [comment]
With Vite under the hood now, this seems like it could be a real option.
